preprocess/analysis/language_analysis.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Jun 23 16:07:17 2020
+
+@author: robert
+"""
+
+# Imports the Google Cloud client library
+from google.cloud import language
+from google.cloud.language import enums
+from google.cloud.language import types
+
+import json
+
+# See readme about how to setup mykey.json
+CLIENT = language.LanguageServiceClient.from_service_account_json("mykey.json")
+ENCODING_TYPE = enums.EncodingType.UTF8
+
+
+def round_to_1(number):
+  """Rounds number to the nearest tenth. 
+
+  Cloud language api will automatically return floats to like 15 decimal 
+  places. It's convenient to just round down. 
+
+  Args: 
+    number: float or an int.
+  
+  Returns:
+    a float to tens precisions.
+  """
+  
+  return float(round(number, 1))
+
+def round_to_2(number): 
+  """Rounds number to the nearest hundreth.
+  
+  Args: 
+    number: float or an int.
+  
+  Returns:
+    a float to hundreds precisions.
+  """
+  
+  return float(round(number, 2))
+
+
+
+def analyze_sentiment(text):
+  """Applies sentiment analysis to an input string.
+  
+  Calls cloud language API once for sentiment analysis if input is a string. 
+  Rounds sentiment and magnitude to the nearest tenth for convenience. 
+
+  Args: 
+    text: an input string.
+  
+  Returns: 
+    A json string mapping score and magnitude keys to their corresponding
+    value.
+    
+  Raises:
+    TypeError: input is not a string.
+  """
+  
+  if not isinstance(text, str): 
```

Comment:
@rob-marcus to check for empty string before isinstance, you can have a condition of `if not (str and isinstance(text, str)):`

Reply:
Empty strings are valid inputs technically and will just return '{"score": 0.0, "magnitude": 0.0}' or []. 

The same goes for whitespace strings, i.e., `"\n\n\n\n\n"`, `" "*1000`, etc.  

Since these are acceptable inputs and outputs, I'm going to say they definitely aren't a `TypeError`. 

I think it's debatable if they are another type of error. If an ads blank/empty, I don't think the place to throw an error or not process that ad is in sentiment analysis. Probably somewhere else in the processing/parsing stream.

Reply:
The api will also handle non-`utf-8` encodings.